### PR TITLE
[gateway] enable to restrict shown buckets 

### PR DESCRIPTION
**Pull request by mizuno-keyence**
## Description

- Added a `--bucket-filter` option to  gateway sub command that restricts shown buckets on the object browser.
- Fixed a bug that gateway command flag registration runs into an error .
    - the same flags are registered multiple times.

## Motivation and Context

Allow users who can access MinIO browser to see and download subset of objects on the GCS without authentication to a Google Cloud.

## Regression
No

## How Has This Been Tested?

- Added a unit test against bucket filtering.
- Run GCS gateway with bucket-filter option and checked that only buckets whose name matches the given pattern are shown on a browser.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project. - Applied `gofmt`
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#7d3b9973-b7d0-4364-9f89-527acca8322b) |
| mint-dist-xl.sh | :x: [more...](#dd8d83d6-c2ba-42f1-9ddd-b28a1ffd63b6) |
<a name="7d3b9973-b7d0-4364-9f89-527acca8322b"></a>
### 7819-c7ed42f/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31042
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp eb4eafe68fd4:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 48 seconds
(4/14) Running aws-sdk-ruby tests ... done in 14 seconds
(5/14) Running awscli tests ... done in 2 minutes and 4 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... FAILED in 2 minutes and 51 seconds
{
  "name": "mc",
  "duration": "130417",
  "function": "test_mirror_list_objects_storage_class",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-19482 --quiet --no-color mirror --storage-class REDUCED_REDUNDANCY /mint/data myminio/mc-test-bucket-9187  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-9187/datafile-1-MB`\n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-9187/datafile-0-b`\n`/mint/data/datafile-1-kB` -> `myminio/mc-test-bucket-9187/datafile-1-kB`\n`/mint/data/datafile-1-b` -> `myminio/mc-test-bucket-9187/datafile-1-b`\n`/mint/data/datafile-100-kB` -> `myminio/mc-test-bucket-9187/datafile-100-kB`\n`/mint/data/datafile-10-MB` -> `myminio/mc-test-bucket-9187/datafile-10-MB`\n`/mint/data/datafile-129-MB` -> `myminio/mc-test-bucket-9187/datafile-129-MB`\n`/mint/data/datafile-1.03-MB` -> `myminio/mc-test-bucket-9187/datafile-1.03-MB`\n`/mint/data/datafile-11-MB` -> `myminio/mc-test-bucket-9187/datafile-11-MB`\n`/mint/data/datafile-10-kB` -> `myminio/mc-test-bucket-9187/datafile-10-kB`\n`/mint/data/datafile-33-kB` -> `myminio/mc-test-bucket-9187/datafile-33-kB`\n`/mint/data/datafile-5243880-b` -> `myminio/mc-test-bucket-9187/datafile-5243880-b`\n`/mint/data/datafile-5-MB` -> `myminio/mc-test-bucket-9187/datafile-5-MB`\n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-9187/datafile-65-MB`\n`/mint/data/datafile-6-MB` -> `myminio/mc-test-bucket-9187/datafile-6-MB`\nmc: <ERROR> Failed to copy `/mint/data/datafile-129-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-11-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-65-MB`. Please reduce your request\nTotal: 233.17 MiB, Transferred: 565.53 MiB, Speed: 4.34 MiB/s"
}

Executed 6 out of 14 tests successfully.
```
<a name="dd8d83d6-c2ba-42f1-9ddd-b28a1ffd63b6"></a>
### 7819-c7ed42f/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31020
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 677311ebf6cd:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 1 minutes and 53 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 3 seconds
(8/14) Running minio-dotnet tests ... done in 34 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 2 seconds
(10/14) Running minio-java tests ... FAILED in 36 seconds
{
  "name": "minio-java",
  "function": "listIncompleteUploads(final String bucketName, final String prefix, final boolean recursive)",
  "args": "prefix: minio, recursive: true",
  "duration": 296,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code=SlowDown, message=Please reduce your request, bucketName=minio-java-test-3vlohda, objectName=minio-java-test-22qqld0, resource=/minio-java-test-3vlohda/minio-java-test-22qqld0, requestId=15AA31B46FD80015, hostId=ad08185d-f4f8-4df4-ae58-1d408d4825dd)\nrequest={method=DELETE, url=http://72.28.97.60:31020/minio-java-test-3vlohda/minio-java-test-22qqld0?uploadId=02b4e543-16bc-4a05-a85d-1d2a6e79e8ba, headers=Host: 72.28.97.60:31020\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20190621T110111Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20190621/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=503, headers=Accept-Ranges: bytes\nContent-Length: 363\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nRetry-After: 120\nServer: MinIO/DEVELOPMENT.2019-06-21T10-42-10Z\nVary: Origin\nX-Amz-Request-Id: 15AA31B46FD80015\nX-Minio-Deployment-Id: ad08185d-f4f8-4df4-ae58-1d408d4825dd\nX-Xss-Protection: 1; mode=block\nDate: Fri, 21 Jun 2019 11:01:11 GMT\n}\n >>> [io.minio.MinioClient.executeReq(MinioClient.java:1204), io.minio.MinioClient.execute(MinioClient.java:1070), io.minio.MinioClient.executeDelete(MinioClient.java:1347), io.minio.MinioClient.abortMultipartUpload(MinioClient.java:5149), io.minio.MinioClient.putObject(MinioClient.java:4337), io.minio.MinioClient.putObject(MinioClient.java:4183), FunctionalTest.listIncompleteUploads_test3(FunctionalTest.java:1679), FunctionalTest.runTests(FunctionalTest.java:2990), FunctionalTest.main(FunctionalTest.java:3118)]"
}

Executed 9 out of 14 tests successfully.
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
We don't really want to do this - please keep a fork of MinIO with this feature.

**Comment by mizuno-keyence** [comment]
Thanks for a comment.
Any reasons you don't want to do that?

**Comment by harshavardhana** [comment]
This increases scope of S3 gateway - which we have no interest right now. We would like to limit the functionalities to keep this part of the code predictable and testable.

**Comment by mizuno-keyence** [comment]
Is it not acceptable adding this only to GCS gateway layer?

If so I reopen PR as what contains only bugfix.
Gateway command can't have any flags due to duplicating `cmd.Flags` registration. 
It seems like a bug, right?

File: cmd/gateway-main.go [hunk 1]
```diff
@@ -37,18 +37,34 @@ func init() {
...
 		Usage:           "start object storage gateway",
-		Flags:           append(ServerFlags, GlobalFlags...),
+		Flags:           append(gatewayFlags, append(ServerFlags, GlobalFlags...)...),
 		HideHelpCommand: true,
 	}
 )
 
 // RegisterGatewayCommand registers a new command for gateway.
 func RegisterGatewayCommand(cmd cli.Command) error {
-	cmd.Flags = append(append(cmd.Flags, append(cmd.Flags, ServerFlags...)...), GlobalFlags...)
```
Comment (mizuno-keyence) [inline cmd/gateway-main.go:51]: This appends `cmd.Flags` twice so that we can't have any flags at each gateway command.
(It's blocked by an error)

**Pull request closed without merging**